aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -0,0 +1,511 @@
+        } else {
+          // collect weight at max/min indices
+          index_t col = col_data[arg_out_data[m * K + k]];
+          grad_ptr[k] = other_data[col * K + k] * grad_out_ptr[k];
+        }
+      }
+    }
+  });
+
+  // scatter_add, consider to parallel this with atomic
+  for (const auto i : c10::irange(M * K)) {
+    index_t ind = arg_out_data[i];
+    if (ind != index_t(nnz)) {
+      grad_values_data[ind] += grad_data[i];
+    }
+  }
+}
+
+template <typename scalar_t, typename index_t>
+void spmm_reduce_update_values_kernel_impl(
+    const Tensor& updated_values,
+    const Tensor& values,
+    const Tensor& crow_indices,
+    const Tensor& row_indices) {
+
+  int64_t nnz = values.numel();
+  if (nnz == 0) {
+    return;
+  }
+
+  auto updated_values_data = updated_values.accessor<scalar_t, 1>();
+  auto values_data = values.accessor<scalar_t, 1>();
+  auto crow_data = crow_indices.accessor<index_t, 1>();
+  auto row_data = row_indices.accessor<index_t, 1>();
+
+  at::parallel_for(0, nnz, 1, [&](int64_t begin, int64_t end) {
+    for (const auto i : c10::irange(begin, end)) {
+      index_t row = row_data[i];
+      index_t row_start = crow_data[row], row_end = crow_data[row + 1];
+      updated_values_data[i] = values_data[i] / std::max((index_t)1, row_end - row_start);
+    }
+  });
+}
+
+template <typename scalar_t, typename index_t>
+void spmm_reduce_backward_other_arg_kernel_impl(
+    const Tensor& grad_other,
+    const Tensor& grad_out_,
+    const Tensor& col_indices,
+    const Tensor& values,
+    const Tensor& arg_out_) {
+
+  int64_t nnz = values.numel();
+  if (nnz == 0) {
+    return;
+  }
+
+  auto grad_out = grad_out_.contiguous();
+  auto arg_out = arg_out_.contiguous();
+
+  scalar_t* grad_other_data = grad_other.data_ptr<scalar_t>();
+  scalar_t* grad_out_data = grad_out.data_ptr<scalar_t>();
+  auto col_data = col_indices.accessor<index_t, 1>();
+  auto values_data = values.accessor<scalar_t, 1>();
+  index_t* arg_out_data = arg_out.data_ptr<index_t>();
+
+  int64_t M = grad_out.size(0);
+  int64_t K = grad_out.size(1);
+  auto grad = at::empty({M, K}, grad_out.options());
+  scalar_t* grad_data = grad.data_ptr<scalar_t>();
+
+  at::parallel_for(0, M, 1, [&](int64_t begin, int64_t end) {
+    for (const auto m : c10::irange(begin, end)) {
+      scalar_t* grad_out_ptr = grad_out_data + m * K;
+      scalar_t* grad_ptr = grad_data + m * K;
+      index_t* arg_out_ptr = arg_out_data + m * K;
+
+      for (const auto k : c10::irange(K)) {
+        if (arg_out_ptr[k] == index_t(nnz)) {
+          grad_ptr[k] = scalar_t(0);
+        } else {
+          grad_ptr[k] = values_data[arg_out_ptr[k]] * grad_out_ptr[k];
+        }
+      }
+    }
+  });
+
+  // scatter_add, consider to parallel this with atomic
+  for (const auto m : c10::irange(M)) {
+    for (const auto k : c10::irange(K)) {
+      index_t ind = arg_out_data[m * K + k];
+      if (ind != index_t(nnz)) {
+        index_t col = col_data[ind];
```

Comment:
`ind` records the col index of the max or min value, which is used is backward. since `col` fro CSR tensor is length of `nnz`, then the col index would be `0<= ind < nnz`.

`arg_out` is filled with `nnz` at  init: `arg_out.fill_(nnz);` that indicats is `ind != nnz`, then the element has been updated in the forward.